### DISCUSSION TITLE
Welcome to CodeWorld Discussions!

### DISCUSSION THREAD

**Discussion by cdsmith** [google/codeworld#1473] [General]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

**Comment by smmohash** [comment]
I am an assistant of Prof. Voigtländer of the University Duisburg-Essen in Germany, and I would like to use CodeWorld for educational purposes (Programming Paradigms : Haskell).
I have been trying to install CodeWorld on a virtual machine but having no success. I would like some help if that is an ok thing to ask for. Thank you in advance.

**Reply by cdsmith** [reply]
Sure, I'd love to help.  Can you tell me more about what you mean by "having no success"?  I believe you ought to be able to just run `install.sh` as-is on a Debian or Ubuntu virtual machine.  It's been longer since I've tried any other Linux distributions, so there might be some work there.  I don't know that it works at all on anything non-Linux.

The initial install takes a long time, since it sets up its own Haskell setup, including a custom build of GHC, GHCJS, all system libraries, and a whole slew of other libraries.  You can pare down the last by editing codeworld-available-packages, but you should still expect an hour or two.

If you're still having trouble, please let me know the details of the OS and version, and the error message you see.

**Reply by cdsmith** [reply]
By the way, you're also welcome to direct your students to the public hosted server at https://code.world/haskell, without installing it locally.  The only reason to install locally is if you want to change something.

**Reply by smmohash** [reply]
Thanks for the kind reply. 
My operating system on my virtual machine is Linux Mint (So, Ubuntu). Allocated is: 30 to 35 GB of space and 6 GB or RAM.
And yes, we would like to modify CodeWorld.

The following happens when I run `install.sh`.
Note that I had to change `run . nodejs build/bin/find-dup-modules.jsexe/all.js \` to `run . /usr/bin/node build/bin/find-dup-modules.jsexe/all.js \`

![2](https://user-images.githubusercontent.com/33121093/125511800-a39ffa65-63ba-4212-afc9-deea5f295fa2.png)
![1](https://user-images.githubusercontent.com/33121093/125511805-2faad29f-77f8-4b45-966d-c14aa4b232b5.png)

**Reply by cdsmith** [reply]
Oh, interesting.  Looks like you got two different versions of Cabal installed.  You can run:

```
(. base.sh && ghcjs-pkg hide Cabal-2.4.0.1)
```

and that should fix it.  I'll look into how this happened.

**Reply by smmohash** [reply]
Thank you again. I did what you said, and I got this:

![image](https://user-images.githubusercontent.com/33121093/125705266-df2ac340-d683-4755-80fa-683c4179b758.png)

I have not installed two different versions of Cabal myself. But it seems to have happened somehow anyway.
All I did was:
- istalling `nodejs`
- running `install.sh`

**Reply by cdsmith** [reply]
Okay, you can fix this with:

```
(. base.sh && ghcjs-pkg hide cabal-doctest-1.0.7)
```

**Reply by smmohash** [reply]
I have reset my virtual machine. Could you please tell me what steps to follow?
Or are you still looking into how this happened?

**Reply by cdsmith** [reply]
I'm still not sure how it happened.  When you get this error, it's because the same module is provided by two different packages, and you need to use ghcjs-pkg to hide one of them.  You can do that manually for now.  Once I figure out exactly what happened, I can add the commands to the build script to do so automatically.

**Reply by smmohash** [reply]
I reached this point after applying the fix.:
![image](https://user-images.githubusercontent.com/33121093/126543958-d9bd488f-30cb-4e0e-b74e-c3f1df3ba965.png)

Now I have this:
![new problem](https://user-images.githubusercontent.com/33121093/126544054-5f4ff268-77f9-494d-b709-250b2ec214f3.PNG)

**Reply by smmohash** [reply]
Could you help me please?

**Reply by smmohash** [reply]
This seemed to have solved the issue. It finished executing.

```diff
diff --git a/build.sh b/build.sh
index 87bf3de0..e87019e2 100755
--- a/build.sh
+++ b/build.sh
@@ -39,12 +39,10 @@ run . ghcjs-pkg hide simple-affine-space     # Conflicts with vector-space
 run . ghcjs-pkg hide newtype                 # Replaced by newtype-generics
 run . ghcjs-pkg hide non-empty               # Conflicts with semialign
 run . ghcjs-pkg hide hgeometry-combinatorial # Conflicts with random-shuffle
-(. base.sh && ghcjs-pkg hide Cabal-2.4.0.1)
-(. base.sh && ghcjs-pkg hide cabal-doctest-1.0.7)
 
 # Check for duplicate modules.  Fail the build if so, since that's a
 # poor user experience.
-run . /usr/bin/node build/bin/find-dup-modules.jsexe/all.js \
+run . nodejs build/bin/find-dup-modules.jsexe/all.js \
              ~/.ghcjs/x86_64-linux-8.6.0.1-8.6.5/ghcjs/package.conf.d/package.cache
 
 run codeworld-base  cabal configure --ghcjs
@@ -78,7 +76,7 @@ run .  cabal_install --ghcjs ./funblocks-client
 # Build the CodeMirror JavaScript bundle.
 function build_codemirror {
   node_modules/uglify-js/bin/uglifyjs \
-      codemirror-compressed.js \
+      lib/codemirror.js \
       addon/dialog/dialog.js \
       addon/display/placeholder.js \
       addon/display/rulers.js \
```

It finished executing without any errors.
But after I execute `run.sh` and open my internet browser (http://localhost:8080), I get the text field where I can type in my code as well as the run button, but clicking it does not do anything.

![image](https://user-images.githubusercontent.com/33121093/129394034-019cefd0-470f-499b-a0a2-a24ef493e4a0.png)

I believe this is similar to https://github.com/google/codeworld/issues/1496

**Reply by cdsmith** [reply]
You're seeing codemirror failing to initialize correctly.  That's not an unexpected result of changing the codemirror build, if you've forgotten parts that are needed.

I'm retrying the install right now.  If I can reproduce the problem, I'll dig into it further.  If not, I might need more specific info from you about operating system versions or configuration.  I'll let you know.

**Reply by cdsmith** [reply]
I was able to reproduce and committed a fix for the errors from `find-dup-modules` (see commit a27a7c6).  However, I cannot reproduce the problem with CodeMirror.  I'm also confused by the diff you posted above.  Did you run diff backwards?  It looks like you're showing a diff for reverting back to the checked-in config.

The error you're getting suggests that you've got a bad checkout of CodeMirror.  Not sure how that happened, but I'd try the following:

1. Revert your local changes (`git reset --hard`) and then `git pull`.
2. Run `rm build/progress/codemirror`
3. Rerun `./install.sh`

That should delete your local branch of codemirror and refetch it from the remote repository.  If your local checkout was corrupted somehow, this will fix it.

**Reply by smmohash** [reply]
Yes, I ran the diff in reverse. I am sorry about that.

These are the changes that I applied:

```
diff --git a/build.sh b/build.sh
index ca05e138..e78e8da2 100755
--- a/build.sh
+++ b/build.sh
@@ -44,7 +44,7 @@ run . ghcjs-pkg hide cabal-doctest           # Conflicts between multiple versio
 
 # Check for duplicate modules.  Fail the build if so, since that's a
 # poor user experience.
-run . nodejs build/bin/find-dup-modules.jsexe/all.js \
+run . /usr/bin/node build/bin/find-dup-modules.jsexe/all.js \
              ~/.ghcjs/x86_64-linux-8.6.0.1-8.6.5/ghcjs/package.conf.d/package.cache
 
 run codeworld-base  cabal configure --ghcjs
@@ -78,7 +78,7 @@ run .  cabal_install --ghcjs ./funblocks-client
 # Build the CodeMirror JavaScript bundle.
 function build_codemirror {
   node_modules/uglify-js/bin/uglifyjs \
-      lib/codemirror.js \
+      codemirror-compressed.js \
       addon/dialog/dialog.js \
       addon/display/placeholder.js \
       addon/display/rulers.js \
```

With these changes, I do not get any error messages when I execute `build.sh` , but it still does not work (as shown in the screenshot above of my internet browser).

I could not successfully complete the build without these changes.
Could you please tell me what your operating system exactly is (which version)?

**Reply by cdsmith** [reply]
Hi.  The problem you're running into with CodeMirror is caused by your change to the `build.sh` script when you build the compressed CodeMirror file.  `codemirror-compressed.js` should be the output from that command, not an input.  You are using it as both, which probably means there's something more or less random in that file now.  You need to revert that change.

This leaves the question of why you were missing `lib/codemirror.js`.  One possibility is that you never finished running `install.sh`, which builds that file among other things.  I suggest that you revert that change, then run `rm build/progress/codemirror`, and then rerun `install.sh`.  This will reinstall codemirror, including building `lib/codemirror.js` again.

To answer your last question, the production CodeWorld server is currently using Debian GNU/Linux 9.4 (stretch)
